wicket/src/update.rs

```diff
@@ -0,0 +1,241 @@
+
+#[derive(Debug, Clone)]
+pub struct Sha3_256Digest([u8; 32]);
+
+impl Serialize for Sha3_256Digest {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        hex::serde::serialize(self.0, serializer)
+    }
+}
+
+impl<'de> Deserialize<'de> for Sha3_256Digest {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        Ok(Sha3_256Digest(hex::serde::deserialize(deserializer)?))
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub enum ArtifactType {
+    // Sled Artifacts
+    SledSp,
+    SledRoT,
+    HostPhase1,
+    HostPhase2,
+
+    // PSC Artifacts
+    PscSp,
+    PscRot,
+
+    // Switch Artifacts
+    SwitchSp,
+    SwitchRot,
+}
+
+/// A description of a software artifact that can be installed on the rack
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Artifact {
+    filename: PathBuf,
+    artifact_type: ArtifactType,
+    version: Version,
+    digest: Sha3_256Digest,
+    length: u64,
+}
+
+/// Attempt to convert an ArtifactSpec to an Artifact
+/// by reading from the filesysetm and hashing.
+impl TryFrom<ArtifactSpec> for Artifact {
+    type Error = std::io::Error;
+    fn try_from(spec: ArtifactSpec) -> Result<Self, Self::Error> {
+        let mut hasher = Sha3_256::new();
+        let mut file = File::open(&spec.filename)?;
+        let length = std::io::copy(&mut file, &mut hasher)?;
+        let mut digest = [0u8; 32];
+        digest.copy_from_slice(&hasher.finalize());
+        let digest = Sha3_256Digest(digest);
+        Ok(Artifact {
+            filename: spec.filename,
+            artifact_type: spec.artifact_type,
+            version: spec.version,
+            digest,
+            length,
+        })
+    }
+}
+
+/// User Input that describes artifacts as part of the [`RackUpdateSpec`]
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct ArtifactSpec {
+    filename: PathBuf,
+    artifact_type: ArtifactType,
+    version: Version,
+}
+
+/// The set of all artifacts in an update
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Manifest {
+    version: Version,
+    artifacts: Vec<Artifact>,
+}
+
+impl Manifest {
+    pub fn dump<P: AsRef<Path>>(
+        &self,
+        path: P,
+    ) -> Result<PathBuf, std::io::Error> {
+        let json = serde_json::to_string(&self)?;
+        let path = path.as_ref().join("manifest.json");
+        let mut file = File::create(&path)?;
+        file.write_all(json.as_bytes())?;
+        Ok(path)
+    }
+
+    /// Unpack a Tar archive into `unpack_dir`, read the manifest, and return
+    /// the manifest.
+    pub fn load(
+        tarfile: impl AsRef<Path>,
+        unpack_dir: impl AsRef<Path>,
+    ) -> std::io::Result<Manifest> {
+        let tarfile = File::open(tarfile)?;
+        let mut archive = tar::Archive::new(tarfile);
+        archive.unpack(&unpack_dir)?;
```

Comment:
This unpacks the entire archive, which is potentially quite large, right? I realize there is more work to come to use the contents of the archive, but a couple thoughts:

* Should we return some kind of handle that contains both the `Manifest` and `unpack_dir`, so we don't have to require our caller to pass the same `unpack_dir` to later operations?
* Or, should we leave the unpacking / dir management to our caller entirely, and require that _they_ do the unpacking and only pass us a path to a directory containing an already-unpacked tarball?

Also fine punting on this until more code shows up that wants to use the unpacked tarball and revisiting then.

Reply:
Yeah, I knew you were gonna comment on this one. Of all the things in this PR, I liked this the least. I'll think a bit more about how to resolve this.

Reply:
I'm gonna punt on this for now, and think about it more once we start using things. I expect this all to change a bit with version trees and such anyway.